datumaro/components/project.py

```diff
@@ -2,486 +2,2184 @@
+            config = TreeConfig(config)
+        if config.format_version != 2:
+            raise ValueError("Unexpected tree config version '%s', expected 2" %
+                config.format_version)
+        self._config = config
+
+        self._project = project
+        self._rev = rev
+
+        self._sources = ProjectSources(self)
+        self._targets = ProjectBuildTargets(self)
+
+    def save(self):
+        self.dump(self._config.config_path)
+
+    def dump(self, path):
+        os.makedirs(osp.dirname(path), exist_ok=True)
+        self._config.dump(path)
+
+    @property
+    def sources(self) -> ProjectSources:
         return self._sources
 
-    def _save_branch_project(self, extractor, save_dir=None):
-        if not isinstance(extractor, Dataset):
-            extractor = Dataset.from_extractors(
-                extractor) # apply lazy transforms to avoid repeating traversals
+    @property
+    def build_targets(self) -> ProjectBuildTargets:
+        return self._targets
+
+    @property
+    def config(self) -> Config:
+        return self._config
+
+    @property
+    def env(self) -> Environment:
+        return self._project.env
+
+    @property
+    def rev(self) -> str:
+        return self._rev
+
+    def make_dataset(self, target: Optional[str] = None) -> Dataset:
+        if not target:
+            target = 'project'
+
+        pipeline = self.build_targets.make_pipeline(target)
+        return ProjectBuilder(self._project, self).make_dataset(pipeline)
+
+    @property
+    def is_working_tree(self) -> bool:
+        return not self._rev
+
+    def source_data_dir(self, source) -> str:
+        if self.is_working_tree:
+            return self._project.source_data_dir(source)
+
+        obj_hash = self.build_targets[source].head.hash
+        return self._project.cache_path(obj_hash)
+
+
+class DiffStatus(Enum):
+    added = auto()
+    modified = auto()
+    removed = auto()
+    missing = auto()
+    foreign_modified = auto()
+
+Revision = NewType('Revision', str) # a commit hash or a named reference
+Reference = NewType('Reference', str) # a commit or an object hash
+
+class Project:
+    @staticmethod
+    def find_project_dir(path: str) -> Optional[str]:
+        path = osp.abspath(path)
+
+        if path.endswith(ProjectLayout.aux_dir) and osp.isdir(path):
+            return path
 
-        # NOTE: probably this function should be in the ViewModel layer
-        save_dir = osp.abspath(save_dir)
-        if save_dir:
-            dst_project = Project()
+        temp_path = osp.join(path, ProjectLayout.aux_dir)
+        if osp.isdir(temp_path):
+            return temp_path
+
+        return None
+
+    def _migrate_from_v1_to_v2(self) -> bool:
+        old_config_path = osp.join(self._aux_dir, 'config.yaml')
+        old_config = Config.parse(old_config_path)
+        if old_config.format_version == 1:
+            pass
+        elif old_config.format_version == 2:
+            return
```

Comment:
This was supposed to avoid occasional config filename clashes. Currently it should not happen, yes.

Reply:
> This was supposed to avoid occasional config filename clashes. Currently it should not happen, yes. But the difference is quite little - config.yaml vs config.yml.

I don't understand what your point is. No matter how small, there is still a difference. All new config-handling code expects to see `config.yml`, so the upgrade code should either create `config.yml` or indicate an error if that's not possible.

Reply:
> My point is that this difference is occasional. It is not a system requirement.

What does this mean and why does it matter? The difference exists. Therefore, the new project code will not work with a config named `config.yaml`. Therefore, `_migrate_from_v1_to_v2` must either create a `config.yml` or raise.